metricbeat/module/system/filesystem/helper.go

```diff
@@ -112,8 +115,10 @@ func filterFileSystemList(fsList []sigar.FileSystem) []sigar.FileSystem {
 // GetFileSystemStat retreves stats for a single filesystem
 func GetFileSystemStat(fs sigar.FileSystem) (*FSStat, error) {
 	stat := sigar.FileSystemUsage{}
-	if err := stat.Get(fs.DirName); err != nil {
-		return nil, err
+	if fs.SysTypeName != UnavailableDisk {
```

Comment:
Do you have a sample event when this happens? Are stats populated with zero values?

Reply:
no stats json object, the example is right in this ticket description

Reply:
Oh ok, I see. There are all populated with zeroes:
```
    "filesystem": {
      "available": 0,
      "free": 0,
      "used": {
        "pct": 0,
        "bytes": 0
      },
      "type": "unavailable",
      "device_name": "D:\\",
      "mount_point": "D:\\",
      "total": 0
    }
```

This may be confusing, for example when preparing alerts, you may need to use other fields to check if the available space is 0 because the filesystem is actually full or because the metric was not available. I think that in these cases these metrics shouldn't be reported:
```
    "filesystem": {
      "type": "unavailable",
      "device_name": "D:\\",
      "mount_point": "D:\\",
    }
```

The event may still be useful for inventory reasons.

Reply:
somehow I thought the stats would not be there.
I had a look at the function retrieving the stats and it is reused by the `fsstat` metricset. The `fsstat` metricset logs the exception as a debug message and continues to the next disk. I tried to do the same for the `filesystem` for consistency purposes.

So the debug logs will show:
```
Line 46: {"log.level":"debug","@timestamp":"2021-11-15T16:26:13.465+0100","log.logger":"system.fsstat","log.origin":{"file.name":"fsstat/fsstat.go","file.line":85},"message":"error fetching filesystem stats for 'D:\\': GetDiskFreeSpaceEx failed: The device is not ready.","service.name":"metricbeat","ecs.version":"1.6.0"}
Line 47: {"log.level":"debug","@timestamp":"2021-11-15T16:26:13.465+0100","log.logger":"system.filesystem","log.origin":{"file.name":"filesystem/filesystem.go","file.line":85},"message":"error fetching filesystem stats for 'D:\\': GetDiskFreeSpaceEx failed: The device is not ready.","service.name":"metricbeat","ecs.version":"1.6.0"}
```
nothing changed for `fsstat`
but for `filesystem` events will look like this:
```
 {
        "_index" : "metricbeat-8.1.0-2021.11.15-000001",
        "_type" : "_doc",
        "_id" : "mis1JH0BLuBcqFZ9zBxN",
        "_score" : null,
        "_source" : {
          "@timestamp" : "2021-11-15T15:27:24.244Z",
          "system" : {
            "filesystem" : {
              "type" : "ntfs",
              "device_name" : """C:\""",
              "mount_point" : """C:\""",
              "total" : 1004205502464,
              "available" : 477790535680,
              "free" : 477790535680,
              "used" : {
                "bytes" : 526414966784,
                "pct" : 0.5242
              }
            }
          },
          "ecs" : {
            "version" : "8.0.0"
          },
          "host" : {
            "name" : "DESKTOP-K76UDQL"
          },
          "agent" : {
            ...
          },
          "metricset" : {
            "name" : "filesystem",
            "period" : 10000
          },
          "event" : {
            "dataset" : "system.filesystem",
            "module" : "system",
            "duration" : 12886200
          },
          "service" : {
            "type" : "system"
          }
        },
        "sort" : [
          1636990044244
        ]
      },
      {
        "_index" : "metricbeat-8.1.0-2021.11.15-000001",
        "_type" : "_doc",
        "_id" : "nCs1JH0BLuBcqFZ9zBxN",
        "_score" : null,
        "_source" : {
          "@timestamp" : "2021-11-15T15:27:24.244Z",
          "agent" : {
           ...
          },
          "event" : {
            "dataset" : "system.filesystem",
            "module" : "system",
            "duration" : 21079100
          },
          "metricset" : {
            "name" : "filesystem",
            "period" : 10000
          },
          "service" : {
            "type" : "system"
          },
          "system" : {
            "filesystem" : {
              "type" : "unavailable",
              "device_name" : """D:\""",
              "mount_point" : """D:\"""
            }
          },
          "ecs" : {
            "version" : "8.0.0"
          },
          "host" : {
            "name" : "DESKTOP-K76UDQL"
          }
        },
        "sort" : [
          1636990044244
        ]
      }
```
@fearful-symmetry , @jsoriano wdyt?

Reply:
Agreeing with @jsoriano , we shouldn't report "null zero" metrics. The fields should be omitted where possible, and I think having "bare" events with no/few metrics in this case isn't terrible.